File: platform/android/lynx_android/src/main/java/com/lynx/tasm/animation/keyframe/KeyframeManager.java
// Copyright 2019 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.
package com.lynx.tasm.animation.keyframe;

import android.text.TextUtils;
import android.view.View;
import androidx.annotation.Nullable;
import com.lynx.tasm.animation.AnimationInfo;
import com.lynx.tasm.behavior.PropertyIDConstants;
import com.lynx.tasm.behavior.PropsConstants;
import com.lynx.tasm.behavior.StylesDiffMap;
import com.lynx.tasm.behavior.ui.LynxUI;
import java.lang.ref.WeakReference;
import java.util.HashMap;

// Some acronyms are used as the letters are so long.
// kf is keyframe, opa is opacity, tran is translation or transform
// rot is rotation, sca is scale, bColor is backgroundColor

public class KeyframeManager {
  private WeakReference<LynxUI> mUI;
  private AnimationInfo[] mInfos;
  private HashMap<String, LynxKeyframeAnimator> mAnimators;

  public static boolean hasKeyframeAnimation(StylesDiffMap map) {
    return map.hasKey(PropsConstants.ANIMATION) || map.hasKey(PropertyIDConstants.Animation);
  }

  public KeyframeManager(LynxUI ui) {
    mUI = new WeakReference<>(ui);
    mAnimators = new HashMap<>();
  }

  @Nullable
  LynxUI getUI() {
    return mUI.get();
  }

  @Nullable
  View getView() {
    return getUI().getView();
  }

  public void setAnimations(AnimationInfo[] infos) {
    mInfos = infos;
  }

  public void setAnimation(AnimationInfo info) {
    mInfos = new AnimationInfo[] {info};
  }

  public void notifyAnimationUpdated() {
    if (mInfos == null || (getUI().getHeight() == 0 && getUI().getWidth() == 0)) {
      return;
    }
    HashMap<String, LynxKeyframeAnimator> animators = new HashMap<>();
    for (AnimationInfo info : mInfos) {
      if (info == null || TextUtils.isEmpty(info.getName())) {
        continue;
      }
      LynxKeyframeAnimator animator = (mAnimators != null ? mAnimators.get(info.getName()) : null);
      if (animator == null) {
        animator = new LynxKeyframeAnimator(getView(), getUI());
      } else {
        mAnimators.remove(info.getName());
      }
      animators.put(info.getName(), animator);
    }

    // Should destroy all animators that be removed by user firstly, then apply info to other
    // animators.
    if (mAnimators != null) {
      for (LynxKeyframeAnimator animator : mAnimators.values()) {
        animator.destroy();
      }
    }

    // Should ensure that animators apply info in order
    for (AnimationInfo info : mInfos) {
      if (info == null || TextUtils.isEmpty(info.getName())) {
        continue;
      }
      animators.get(info.getName()).apply(info);
    }

    mAnimators = animators;
  }

  public void endAllAnimation() {
    if (mAnimators == null) {
      return;
    }
    for (LynxKeyframeAnimator animator : mAnimators.values()) {
      animator.destroy();
    }
    mAnimators = null;
    mInfos = null;
  }

  public void notifyPropertyUpdated(String name, Object value) {
    if (mAnimators == null) {
      return;
    }
    for (LynxKeyframeAnimator animator : mAnimators.values()) {
      animator.notifyPropertyUpdated(name, value);
    }
  }

  public boolean hasAnimationRunning() {
    if (mAnimators != null) {
      for (LynxKeyframeAnimator animator : mAnimators.values()) {
        if (animator.isRunning()) {
          return true;
        }
      }
    }
    return false;
  }

  public void onAttach() {
    if (mAnimators == null) {
      return;
    }
    for (LynxKeyframeAnimator animator : mAnimators.values()) {
      animator.onAttach();
    }
  }

  public void detachFromUI() {
    mUI = null;
    if (mAnimators == null) {
      return;
    }
    for (LynxKeyframeAnimator animator : mAnimators.values()) {
      animator.detachFromUI();
    }
  }

  public void attachToUI(LynxUI ui) {
    mUI = new WeakReference<>(ui);
    if (mAnimators == null) {
      return;
    }
    for (LynxKeyframeAnimator animator : mAnimators.values()) {
      animator.attachToUI(ui);
    }
  }

  public void onDetach() {
    if (mAnimators == null) {
      return;
    }
    for (LynxKeyframeAnimator animator : mAnimators.values()) {
      animator.onDetach();
    }
  }

  public void onDestroy() {
    endAllAnimation();
  }
}

File: platform/harmony/lynx_harmony/src/main/cpp/animation/keyframe_manager.cc
// Copyright 2026 The Lynx Authors. All rights reserved.
// Licensed under the Apache License Version 2.0 that can be found in the
// LICENSE file in the root directory of this source tree.

#include "platform/harmony/lynx_harmony/src/main/cpp/animation/keyframe_manager.h"

#include <utility>

#include "platform/harmony/lynx_harmony/src/main/cpp/animation/keyframe_animator.h"
#include "platform/harmony/lynx_harmony/src/main/cpp/ui/ui_base.h"

namespace lynx {
namespace tasm {
namespace harmony {

KeyframeManager::KeyframeManager(UIBase* ui) : ui_(ui) {}

KeyframeManager::~KeyframeManager() { EndAllAnimation(); }

void KeyframeManager::SetAnimations(const std::vector<AnimationInfo>& infos) {
  infos_ = infos;
  NotifyAnimationUpdated();
}

void KeyframeManager::NotifyAnimationUpdated() {
  if (infos_.empty()) {
    if (!ui_) return;
  }

  std::unordered_map<std::string, std::unique_ptr<KeyframeAnimator>>
      new_animators;

  for (const auto& info : infos_) {
    if (info.GetName().empty()) {
      continue;
    }

    auto it = animators_.find(info.GetName());
    if (it != animators_.end()) {
      // Reuse existing animator
      new_animators[info.GetName()] = std::move(it->second);
      animators_.erase(it);
    } else {
      // Create new animator
      auto animator = std::make_unique<KeyframeAnimator>(ui_);
      new_animators[info.GetName()] = std::move(animator);
    }
  }

  animators_ = std::move(new_animators);

  // Apply info to animators
  for (const auto& info : infos_) {
    if (info.GetName().empty()) {
      continue;
    }
    auto it = animators_.find(info.GetName());
    if (it != animators_.end() && it->second) {
      it->second->Apply(info);
    }
  }
}

void KeyframeManager::EndAllAnimation() {
  animators_.clear();
  infos_.clear();
}

bool KeyframeManager::HasAnimationRunning() {
  for (const auto& pair : animators_) {
    if (pair.second->IsRunning()) {
      return true;
    }
  }
  return false;
}

}  // namespace harmony
}  // namespace tasm
}  // namespace lynx
